### ISSUE TITLE
Async await support in Functions

### ISSUE THREAD

**Issue by aligin** [davideicardi/DynamicExpresso#57]
Hello. I've succesfully integrated DynamicExpresso (.NET Core) in my project, but there is a question with the Functions, that returns Task. Is there support for this kind of Expressions?

Example:

    var target = new Interpreter();
    Func<int, Task<bool>> func = async (x) => await _someHttpService.Get(x);
    target.SetFunction("foo", func);
    var parsedExpression = target.Parse("true && foo(1)", parameters));

Currentry I receive Exception

    The binary operator AndAlso is not defined for the types 'System.Threading.Tasks.Task`1[System.Boolean]' and 'System.Threading.Tasks.Task`1[System.Boolean]'.

**Comment by davideicardi** [comment]
Yes, currently this is not supported. Probably this can be written using:

> target.Parse("true && await foo(1)", parameters));

Or maybe adding a `ParseAsync` function ... but unfortunately both methods are currently not supported.

The only solution for now is to rewrite your function with sync methods: 

> Func<int, bool> func = (x) => _someHttpService.Get(x).Result;

**Comment by aligin** [comment]
@davideicardi Thanks for the answer. I've already done in sync mode. I realized that the targeted .NET Framework was 4.0. And this version does not contain async await.

**Comment by huangyedie** [comment]
> Yes, currently this is not supported. Probably this can be written using:
> 
> > target.Parse("true && await foo(1)", parameters));
> 
> Or maybe adding a `ParseAsync` function ... but unfortunately both methods are currently not supported.
> 
> The only solution for now is to rewrite your function with sync methods:
> 
> > Func<int, bool> func = (x) => _someHttpService.Get(x).Result;
>
>Is it supported now? Async await

**Comment by metoule** [comment]
No, async / await is not supported, and probably never will be. When you write an async method, the compiler performs a LOT of things (creating a class to hoist captured variables, etc), which can't easily be done. 

Under the hood, we're using Linq expression trees, which doesn't provide any method to do that.

You can't check [this stackoverflow question](https://stackoverflow.com/questions/24240702/can-i-generate-an-async-method-dynamically-using-system-linq-expressions)

**Comment by MarcusGoldschmidt** [comment]
This is a workaround that worked for me

store the results from the async task before executing the invoke of the Lambda

```cs

var interpreter = new Interpreter();

var dictionary = new Dictionary<string, double>();

interpreter.SetFunction("ASYNC_TASK", () => dictionary["ASYNC_TASK"]);

var parsed = interpreter.Parse("ASYNC_TASK() + 1");

// Simulate an async function
dictionary["ASYNC_TASK"] = await Task.FromResult(10);

var result = parsed.Invoke();

Assert.Equal(11.0, result);
```

Also, you can get the Functions that Lambda contains

```cs
List<Identifier> functionsUsedByExpression = parsed
    .Identifiers
    .Where(e => e.GetType().Name == "FunctionIdentifier")
    .ToList();
```
